xls/contrib/integrator/ir_integrator.cc

```diff
@@ -194,6 +244,168 @@ absl::StatusOr<Node*> IntegrationFunction::InsertNode(const Node* to_insert) {
   return inserted;
 }
 
+absl::StatusOr<IntegrationFunction::MergeNodesBackendResult>
+IntegrationFunction::MergeNodesBackend(const Node* node_a, const Node* node_b) {
+  auto validate_node = [this](const Node* node) {
+    if (IntegrationFunctionOwnsNode(node)) {
+      if (!IsMappingTarget(node)) {
+        return absl::InternalError(
+            "Trying to merge non-mapping-target integration node.");
+      }
+    } else {
+      if (HasMapping(node)) {
+        return absl::InternalError(
+            "Trying to merge non-integration node that already has mapping");
+      }
+    }
+    return absl::OkStatus();
+  };
+  XLS_RETURN_IF_ERROR(validate_node(node_a));
+  XLS_RETURN_IF_ERROR(validate_node(node_b));
+  // Simple way to avoid situation where node_a depends on node_b or
+  // vice versa. If we later want to merge nodes from the same function,
+  // we can implement reaching analysis to check for dependencies.
+  if (node_a != node_b) {
+    XLS_RET_CHECK(node_a->function_base() != node_b->function_base());
+  }
+
+  // Separate targets for node_a and node_b because we may derive
+  // map target nodes from the merged node;
+
+  // Identical nodes can always be merged.
+  if (node_a->IsDefinitelyEqualTo(node_b)) {
+    XLS_ASSIGN_OR_RETURN(UnifiedOperands unified_operands,
+                         UnifyNodeOperands(node_a, node_b));
+    XLS_ASSIGN_OR_RETURN(
+        Node * merged,
+        node_a->CloneInNewFunction(unified_operands.operands, function()));
+    MergeNodesBackendResult merge_result;
+    merge_result.can_merge = true;
+    merge_result.target_a = merged;
+    merge_result.target_b = merged;
+    merge_result.added_muxes = unified_operands.added_muxes;
+    merge_result.other_added_nodes.insert(merged);
+    return merge_result;
+  } else {
+    // TODO(jbaileyhandle): Add logic for nodes that are not identical but
+    // may still be merged e.g. different bitwidths for bitwise ops.
+    switch (node_a->op()) {
+      default:
+        break;
+    }
+  }
+
+  MergeNodesBackendResult merge_result;
+  merge_result.can_merge = false;
+  return merge_result;
+}
+
+absl::StatusOr<std::optional<float>> IntegrationFunction::GetMergeNodesCost(
+    const Node* node_a, const Node* node_b) {
+  XLS_ASSIGN_OR_RETURN(MergeNodesBackendResult merge_result,
+                       MergeNodesBackend(node_a, node_b));
+  // Can't merge nodes.
+  if (!merge_result.can_merge) {
+    return absl::nullopt;
+  }
+
+  // Score.
+  float cost = 0.0;
+  auto tabulate_node_elimination_cost = [this, &cost](const Node* node) {
+    if (IntegrationFunctionOwnsNode(node)) {
+      cost -= GetNodeCost(node);
+    }
+  };
+  tabulate_node_elimination_cost(node_a);
+  tabulate_node_elimination_cost(node_b);
+  for (Node* new_node : merge_result.added_muxes) {
+    cost += GetNodeCost(new_node);
+  }
+  for (Node* new_node : merge_result.other_added_nodes) {
+    cost += GetNodeCost(new_node);
```

Comment:
The iteration order here is not deterministic and floating point addition is not associative so this may produce nondeterministic results. Can the cost be an int64? ints have much nicer behavior.

Reply:
Done (although order is now deterministic)